orchagent/bfdorch.cpp

```diff
@@ -0,0 +1,412 @@
+
+    if (multihop)
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_MULTIHOP;
+        attr.value.booldata = true;
+        attrs.emplace_back(attr);
+        fvVector.emplace_back("multihop", "true");
+    }
+    else
+    {
+        fvVector.emplace_back("multihop", "false");
+    }
+
+    if (alias != "default")
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_HW_LOOKUP_VALID;
+        attr.value.booldata = false;
+        attrs.emplace_back(attr);
+
+        Port port;
+        if (!gPortsOrch->getPort(alias, port))
+        {
+            SWSS_LOG_ERROR("Failed to locate port %s", alias.c_str());
+            return false;
+        }
+
+        attr.id = SAI_BFD_SESSION_ATTR_PORT;
+        attr.value.oid = port.m_port_id;
+        attrs.emplace_back(attr);
+
+        attr.id = SAI_BFD_SESSION_ATTR_SRC_MAC_ADDRESS;
+        memcpy(attr.value.mac, port.m_mac.getMac(), sizeof(sai_mac_t));
+        attrs.emplace_back(attr);
+
+        if (!dst_mac_provided)
+        {
+            SWSS_LOG_ERROR("Failed to create BFD session %s: destination MAC address required when hardware lookup not valid",
+                            key.c_str());
+            return false;
+        }
+
+        if (vrf_name != "default")
+        {
+            SWSS_LOG_ERROR("Failed to create BFD session %s: vrf is not supported when hardware lookup not valid",
+                            key.c_str());
+            return false;
+        }
+
+        attr.id = SAI_BFD_SESSION_ATTR_DST_MAC_ADDRESS;
+        memcpy(attr.value.mac, dst_mac.getMac(), sizeof(sai_mac_t));
+        attrs.emplace_back(attr);
+    }
+    else
+    {
+        attr.id = SAI_BFD_SESSION_ATTR_VIRTUAL_ROUTER;
+        if (vrf_name == "default")
+        {
+            attr.value.oid = gVirtualRouterId;
+        }
+        else
+        {
+            VRFOrch* vrf_orch = gDirectory.get<VRFOrch*>();
+            attr.value.oid = vrf_orch->getVRFid(vrf_name);
+        }
+
+        attrs.emplace_back(attr);
+
+        if (dst_mac_provided)
+        {
+            SWSS_LOG_ERROR("Failed to create BFD session %s: destination MAC address not supported when hardware lookup valid",
+                            key.c_str());
+            return false;
+        }
+    }
+
+    fvVector.emplace_back("state", session_state_loopup.at(SAI_BFD_SESSION_STATE_DOWN));
+
+    sai_object_id_t bfd_session_id = SAI_NULL_OBJECT_ID;
+    sai_status_t status = sai_bfd_api->create_bfd_session(&bfd_session_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to create bfd session %s, rv:%d", key.c_str(), status);
+        task_process_status handle_status = handleSaiCreateStatus(SAI_API_BFD, status);
+        if (!parseHandleSaiStatusFailure(handle_status))
+        {
+            return false;
+        }
+    }
+
+    const string state_db_key = vrf_name + state_db_key_delimiter + alias + state_db_key_delimiter + peer_address.to_string();
```

Comment:
Can you move this to a separate function so it can be used from both create and remove functions?

Reply:
Defined a separate function to generate keys.